src/rcl_bindings.cpp

```diff
@@ -1450,6 +1451,648 @@ NAN_METHOD(ServiceServerIsAvailable) {
+      reinterpret_cast<rcl_action_goal_handle_t*>(goal_handle_handle->ptr());
+
+  rcl_action_goal_state_t status;
+  THROW_ERROR_IF_NOT_EQUAL(rcl_action_goal_handle_get_status(
+                               goal_handle, &status),
+                           RCL_RET_OK,
+                           rcl_get_error_string().str);
+
+  v8::Local<v8::Integer> result =
+      Nan::New<v8::Integer>(static_cast<int32_t>(status));
+  info.GetReturnValue().Set(result);
+}
+
+NAN_METHOD(ActionPublishFeedback) {
+  rcl_action_server_t* action_server = reinterpret_cast<rcl_action_server_t*>(
+      RclHandle::Unwrap<RclHandle>(
+          Nan::To<v8::Object>(info[0]).ToLocalChecked())->ptr());
+  void* buffer = node::Buffer::Data(
+      Nan::To<v8::Object>(info[1]).ToLocalChecked());
+
+  THROW_ERROR_IF_NOT_EQUAL(rcl_action_publish_feedback(action_server, buffer),
+                           RCL_RET_OK,
+                           rcl_get_error_string().str);
+
+  info.GetReturnValue().Set(Nan::Undefined());
+}
+
+NAN_METHOD(ActionTakeFeedback) {
+  RclHandle* action_client_handle = RclHandle::Unwrap<RclHandle>(
+      Nan::To<v8::Object>(info[0]).ToLocalChecked());
+  rcl_action_client_t* action_client =
+      reinterpret_cast<rcl_action_client_t*>(action_client_handle->ptr());
+  void* buffer = node::Buffer::Data(
+      Nan::To<v8::Object>(info[1]).ToLocalChecked());
+
+  rcl_ret_t ret = rcl_action_take_feedback(action_client, buffer);
+  if (ret != RCL_RET_OK && ret != RCL_RET_ACTION_CLIENT_TAKE_FAILED) {
+    Nan::ThrowError(rcl_get_error_string().str);
+    rcl_reset_error();
+    info.GetReturnValue().Set(Nan::False());
+    return;
+  }
+
+  if (ret != RCL_RET_ACTION_CLIENT_TAKE_FAILED) {
+    info.GetReturnValue().Set(Nan::True());
+    return;
+  }
+  info.GetReturnValue().Set(Nan::Undefined());
+}
+
+NAN_METHOD(ActionProcessCancelRequest) {
+  RclHandle* action_server_handle = RclHandle::Unwrap<RclHandle>(
+      Nan::To<v8::Object>(info[0]).ToLocalChecked());
+  rcl_action_server_t* action_server = reinterpret_cast<rcl_action_server_t*>(
+      action_server_handle->ptr());
+  void* buffer = node::Buffer::Data(
+      Nan::To<v8::Object>(info[1]).ToLocalChecked());
+  rcl_action_cancel_request_t* cancel_request =
+      reinterpret_cast<rcl_action_cancel_request_t*>(buffer);
+  void* response_buffer = node::Buffer::Data(
+      Nan::To<v8::Object>(info[2]).ToLocalChecked());
+  action_msgs__srv__CancelGoal_Response* response =
+      reinterpret_cast<action_msgs__srv__CancelGoal_Response*>(
+          response_buffer);
+
+  rcl_action_cancel_response_t cancel_response =
+      rcl_action_get_zero_initialized_cancel_response();
+
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+      action_server, cancel_request, &cancel_response);
+  if (ret != RCL_RET_OK) {
+    rcl_action_cancel_response_fini(&cancel_response);
+    Nan::ThrowError(GetErrorMessageAndClear());
+  }
+
+  *response = cancel_response.msg;
```

Comment:
Out of all these changes, this is the one line I was unsure of the most. There's no existing example of how to handle this scenario. The `rcl_action_process_cancel_request` function above expects the response message to be wrapped in a `rcl_action_cancel_response_t` struct meaning I can't pass the response buffer (info[2]) in by reference, so I need to copy it afterwards.

Here's the rclpy usage of this function.

```python
# Cancel response is returned
cancel_response = _rclpy_action.rclpy_action_process_cancel_request(
    self._handle, cancel_request, self._action_type.Impl.CancelGoalService.Response)
```

Here's what I did for rclnodejs:

```javascript
// No return value, cancel response pointer is passed in then deserialized afterwards.
rclnodejs.actionProcessCancelRequest(this.handle, request.serialize(), response.toRawROS());
let cancelResponse = new Response();
cancelResponse.deserialize(response.refObject);
```

This works, but I'm wondering if the `cancel_reponse` memory needs to be cleaned up. If I call `rcl_action_cancel_response_fini(&cancel_response)` after this line it crashes.

Any thoughts on what (if anything) needs to be done here?

Reply:
How about doing this, in cpp:
```cpp
// Allocate struct rcl_action_cancel_response_t and will be destoried in RclHandle::Reset()
rcl_action_cancel_response_t* cancel_response_ptr =  reinterpret_cast<cancel_response_ptr *>(malloc(sizeof(cancel_response_ptr)));
*cancel_response_ptr  = cancel_response;
auto js_obj = RclHandle::NewInstance(cancel_response_ptr , nullptr, [cancel_response_ptr] {
    rcl_action_cancel_response_fini(cancel_response_ptr );
});
info.GetReturnValue().Set(js_obj);
```

and in JavaScript

```js
let msg = rclnodejs.actionProcessCancelRequest(this.handle, request.serialize(), response.toRawROS());
let cancelResponse = new Response();
cancelResponse.deserialize(response.refObject);
msg.release();  // this triggers RclHandle::Reset() and will call rcl_action_cancel_response_fini()
```

I haven't tested this, just a potential method.

Reply:
Thanks. I'll play around with it and let you know when everything is ready.

Reply:
This helped me figure out what to do. Thanks. Will probably close the PR in favor of a different one with e2e action support.